### PR TITLE
HTML Transformation to NITF

### PR DESCRIPTION
**Pull request by hosamaly**
The main class in this PR is `NitfValidator`, the companion object of which has logic to transform existing NITF files (which typically contain the HTML `body` of articles) to become valid NITF files.

The transformations should make it to the main source code shortly to be used after cleaning the HTML using Jsoup. They should help guarantee that we produce valid NITF (according to Amazon's schema).

This test was run against ~27k files generated between 18/08/2017 and 03/04/2018.

### CODE REVIEW

File: src/test/scala/com/gu/nitf/NitfValidator.scala [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+       - stray <li> in 20180117.0111.moved/085_leave-campaigns-350m-claim-was-too-low-says-boris-johnson.nitf
+       - XML entity ("Information Separator 3" control char!) in 20180116.0111.moved/060_hm-stores-in-south-afrtica-trashed-by-protesters-after-racist-ad.nitf
+       - empty <ul> in 20180223.0111.moved/073_freehold-on-disputed-birmingham-leasehold-flats-goes-on-sale.nitf
+       - <ul> with <p> children in 20180217.0111.moved/024_barry-bennell-abuse-manchester-city-crewe.nitf
+       - <li> without <ul> or <ol> in
+         - 20180308.0111.moved/017_disgraced-ex-co-op-bank-boss-paul-flowers-crystal-methodist-banned-from-financial-services.nitf
+         - 20180302.0111.moved/010_judges-told-to-limit-observers-if-witness-has-to-remove-veil.nitf
+  */
+
+  private val basePath = Paths.get("../kindle-publications-extracted/feeds").toRealPath()
```
Comment (hosamaly) [inline src/test/scala/com/gu/nitf/NitfValidator.scala]: I intend to add the test files to the repository (or to S3?) in another commit later.

File: src/test/scala/com/gu/nitf/NitfValidator.scala [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+            case e: org.xml.sax.SAXParseException =>
+              cancel("XML file is invalid!", e)
+          }
+        }
+      }
+    }
+
+  private def validateFile(nitfFile: File): Unit = {
+    val xml = transform(Utility.trim(XML.loadFile(nitfFile)))
+    validateXml(xml, "kpp-nitf-3.5.7.xsd")
```
Comment (regiskuckaertz) [inline src/test/scala/com/gu/nitf/NitfValidator.scala]: is this version going to change over time?

File: src/test/scala/com/gu/nitf/NitfValidator.scala [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+    removeUnsupportedAttributes,
+    unwrapTables,
+    wrapBlockContentText,
+    wrapElements
+  )
+
+  private val setVersionToNitf35 = rewriteRule("set version to NITF 3.5") {
+    case e: Elem if e.label == "nitf" =>
+      e.copy(scope = NamespaceBinding(null, "http://iptc.org/std/NITF/2006-10-18/", TopScope))
+        .withAttribute(e.prefix, "version", "-//IPTC//DTD NITF 3.5//EN")
```
Comment (regiskuckaertz) [inline src/test/scala/com/gu/nitf/NitfValidator.scala]: I remember seeing these values hardcoded somewhere else: is it worth factoring them out, so you can define `setVersion(version: NitfVersion)` instead?

File: src/test/scala/com/gu/nitf/NitfValidator.scala (hunk 2 above)
Comment (hosamaly) [reply src/test/scala/com/gu/nitf/NitfValidator.scala]: It seems unlikely. The last version was produced in 2012.

File: src/test/scala/com/gu/nitf/NitfValidator.scala (hunk 3 above)
Comment (hosamaly) [reply src/test/scala/com/gu/nitf/NitfValidator.scala]: They were in a different project, specifically [nitf-scala](https://github.com/guardian/nitf-scala).

File: src/test/scala/com/gu/nitf/NitfValidator.scala [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+  private type NodeProcessor = Seq[Node] => Seq[Node]
+  /** Partitions a sequence of nodes according to the ''matcher'' predicate, adapts them as required,
+    * and then combines them again.
+    *
+    * Nodes are divided into contiguous sections of nodes that either match or don't match the predicate.
+    * Each contiguous section is passed to ''adaptMatching'' or ''adaptUnmatching''.
+    */
+  private def adaptPartitions(nodes: Seq[Node],
+                              matches: Node => Boolean,
+                              adaptMatching: Seq[Node] => Seq[Node] = identity,
```
Comment (regiskuckaertz) [inline src/test/scala/com/gu/nitf/NitfValidator.scala]: You could use `NodeProcessor` a few times here.

File: src/test/scala/com/gu/nitf/NitfValidator.scala [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+    def adaptUnlessEmpty(partition: Seq[Node], adapt: Seq[Node] => Seq[Node]): Seq[Node] =
+      if (partition.isEmpty) partition else adapt(partition)
+
+    val before = nodes.takeWhile(!matches(_))
+    val middle = nodes.drop(before.size).takeWhile(matches)
+    val after  = nodes.drop(before.size + middle.size)
+
+    adaptUnlessEmpty(before, adaptUnmatching) ++
+      adaptUnlessEmpty(middle, adaptMatching) ++
+      adaptUnlessEmpty(after, adaptPartitions(_, matches, adaptMatching = adaptMatching, adaptUnmatching = adaptUnmatching))
```
Comment (regiskuckaertz) [inline src/test/scala/com/gu/nitf/NitfValidator.scala]: I think it's clear enough that you don't have to bind `adaptMatching` and `adaptUnmatching` by name

File: src/test/scala/com/gu/nitf/NitfValidator.scala [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+    * Each contiguous section is passed to ''adaptMatching'' or ''adaptUnmatching''.
+    */
+  private def adaptPartitions(nodes: Seq[Node],
+                              matches: Node => Boolean,
+                              adaptMatching: Seq[Node] => Seq[Node] = identity,
+                              adaptUnmatching: Seq[Node] => Seq[Node] = identity): Seq[Node] = {
+    def adaptUnlessEmpty(partition: Seq[Node], adapt: Seq[Node] => Seq[Node]): Seq[Node] =
+      if (partition.isEmpty) partition else adapt(partition)
+
+    val before = nodes.takeWhile(!matches(_))
```
Comment (regiskuckaertz) [inline src/test/scala/com/gu/nitf/NitfValidator.scala]: You could save a few trees here 😅 
```scala
val (before, tmp) = nodes.span(!matches(_))
val (middle, after) = tmp.span(matches))
```

File: src/test/scala/com/gu/nitf/NitfValidator.scala [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+  }
+
+  private val convertMisplacedLists = {
+    val isList = (x: Node) => x.label == "ol" || x.label == "ul"
+    val isListItem = (x: Node) => x.label == "li"
+    val nonListItem = (x: Node) => !isListItem(x)
+    rewriteRule("Convert misplaced <li> to <p>") {
+      case e: Elem if isList(e) && e.hasChildren(nonListItem) =>
+        if (e.child.forall(nonListItem))
+          e.child  // remove the extraneous list tag
```
Comment (tomrf1) [inline src/test/scala/com/gu/nitf/NitfValidator.scala]: is this Guardian content containing invalid html?

File: src/test/scala/com/gu/nitf/NitfValidator.scala (hunk 5 above)
Comment (hosamaly) [reply src/test/scala/com/gu/nitf/NitfValidator.scala]: I did that at first, and then spent an hour debugging why it wasn't working, only to find out later that I had the parameters in the wrong order!

File: src/test/scala/com/gu/nitf/NitfValidator.scala (hunk 6 above)
Comment (hosamaly) [reply src/test/scala/com/gu/nitf/NitfValidator.scala]: Nice! Something to learn every day. :)

File: src/test/scala/com/gu/nitf/NitfValidator.scala (hunk 7 above)
Comment (hosamaly) [reply src/test/scala/com/gu/nitf/NitfValidator.scala]: Yes, it is!

File: src/test/scala/com/gu/nitf/NitfValidator.scala [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+        case c => c
+      })
+  }
+
+  private val unwrapTables = rewriteRule("Unwrap tables") {
+    case e: Elem if e.label != "block" && e.hasChildren("table") =>
+      unwrapChildren(e, _.label == "table")
+  }
+
+  private def rewriteRule(ruleName: String)(pf: PartialFunction[Node, Seq[Node]]): RewriteRule = new RewriteRule {
```
Comment (hosamaly) [inline src/test/scala/com/gu/nitf/NitfValidator.scala]: I'm thinking of moving this method and the ones below it to `XmlUtils`.

**Comment by tomrf1** [comment]
👍 nice!

File: src/test/scala/com/gu/nitf/NitfValidator.scala [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+    convertOrRemoveTags,
+    convertMisplacedLists,
+    removeUnsupportedAttributes,
+    wrapBlockContentText,
+    wrapSpecialElements,
+    unwrapBlockContentParents
+  )
+
+  private val removeControlCharacters = {  // temporary rule until https://github.com/scala/scala-xml/pull/203 is fixed
+    val unwanted = (c: Char) => c.isControl && c != '\n' && c != '\r' && c != '\t'
```
Comment (regiskuckaertz) [inline src/test/scala/com/gu/nitf/NitfValidator.scala]: `c.isControl && !c.isWhitespace` 🤗

**Pull request merged**